gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
-        handle->index = (struct gsd_index_entry*)malloc(sizeof(struct gsd_index_entry)
-                                                        * handle->append_index_size);
-        if (handle->index == NULL)
+        total_bytes_written += bytes_written;
+    }
+
+    // sync the expanded index
+    retval = fsync(handle->fd);
+    if (retval != 0)
+    {
+        free(buf);
+        return GSD_ERROR_IO;
+    }
+
+    // free the copy buffer
+    free(buf);
+
+    // update the header
+    handle->header.index_location = new_index_location;
+    handle->file_size = handle->header.index_location + total_bytes_written;
+    handle->header.index_allocated_entries = size_new;
+
+    // write the new header out
+    ssize_t bytes_written
+        = gsd_io_pwrite_retry(handle->fd, &(handle->header), sizeof(struct gsd_header), 0);
+    if (bytes_written != sizeof(struct gsd_header))
+    {
+        return GSD_ERROR_IO;
+    }
+
+    // sync the updated header
+    retval = fsync(handle->fd);
+    if (retval != 0)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    // remap the file index
+    retval = gsd_index_buffer_map(&handle->file_index, handle);
+    if (retval != 0)
+    {
+        return retval;
+    }
+
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Flush the write buffer.
+
+    gsd_write_frame() writes small data chunks into the write buffer. It adds index entries for
+    these chunks to gsd_handle::buffer_index with locations offset from the start of the write
+    buffer. gsd_flush_write_buffer() writes the buffer to the end of the file, moves the index
+    entries to gsd_handle::frame_index and updates the location to reference the beginning of the
+    file.
+
+    @param handle Handle to flush the write buffer.
+    @returns GSD_SUCCESS on success or GSD_* error codes on error
+*/
+static int gsd_flush_write_buffer(struct gsd_handle* handle)
+{
+    if (handle == NULL)
+    {
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (handle->write_buffer.size == 0)
+    {
+        // nothing to do
+        return GSD_SUCCESS;
+    }
+
+    if (handle->buffer_index.size == 0)
+    {
+        // error: bytes in buffer, but no index for them
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+    // write the buffer to the end of the file
+    uint64_t offset = handle->file_size;
+    ssize_t bytes_written = gsd_io_pwrite_retry(handle->fd,
+                                                handle->write_buffer.data,
+                                                handle->write_buffer.size,
+                                                offset);
+
+    if (bytes_written == -1 || bytes_written != handle->write_buffer.size)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    handle->file_size += handle->write_buffer.size;
+
+    // reset write_buffer for new data
+    handle->write_buffer.size = 0;
+
+    // move buffer_index entries to file_index
+    size_t i;
+    for (i = 0; i < handle->buffer_index.size; i++)
```

Comment:
Just declare `i` in the loop since it's not needed after.

Reply:
I would, but some older C compilers throw errors.

Reply:
Why? Is it specifically for `size_t`?

Reply:
Declaring the loop variable in a for loop is available starting in C99. GCC 4.8 defaults to the C89 standard, which results in [this compiler error](https://dev.azure.com/glotzerlab/gsd/_build/results?buildId=1302&view=logs&j=f1f5612d-6dbb-5da1-cdfa-f96216bc9d2d&t=7d294d6f-716b-5102-8eae-8aeada81840c&l=22). See [this question on stackoverflow](https://stackoverflow.com/questions/1287863/c-for-loop-int-initial-declaration) for more discussion. 

We cannot drop gcc 4.8 support because the pyPI manylinux infrastructure uses gcc 4.8. See the [test builds](https://dev.azure.com/glotzerlab/gsd/_build/results?buildId=1304&view=results) which all fail.

It is trivial to request C99 mode in CMake in a way that works across platforms and compilers, but I am not aware of such a feature in setuptools.